### PR TITLE
[Arm64/Unix] Use portable MP optimized new/alloc

### PR DESCRIPTION
**Pull request by sdmaclea**

### CODE REVIEW

**Comment by sdmaclea** [comment]
@arm64-contrib @jkotas @janvorli This is a big win on MP ARM64.  Total elapsed test time to run 11K tests reduced ~33%.

**Comment by sdmaclea** [comment]
Code is largely copied from `jitinterfacegen.cpp`'s version of `void InitJITHelpers1()`

File: src/vm/arm64/stubs.cpp [hunk 1]
```diff
@@ -1086,6 +1087,31 @@ void JIT_TailCall()
 #if !defined(DACCESS_COMPILE) && !defined(CROSSGEN_COMPILE)
 void InitJITHelpers1()
 {
+    STANDARD_VM_CONTRACT;
+
+    _ASSERTE(g_SystemInfo.dwNumberOfProcessors != 0);
+
+    // TODO-ARM64-NYI !defined(FEATURE_IMPLICIT_TLS) Use helpers optimized w/ inline TLS access...
```
Comment (jkotas) [inline src/vm/arm64/stubs.cpp]: This TODO is not necessary. `FEATURE_IMPLICIT_TLS` is as good as what you can get the handcrafter helpers.

Comment (sdmaclea) [inline src/vm/arm64/stubs.cpp]: The comment is for the Windows !`FEATURE_IMPLICIT_TLS` case.

Comment (sdmaclea) [inline src/vm/arm64/stubs.cpp]: It looks like `FEATURE_IMPLICIT_TLS` is not defined for Windows.  Am I wrong?

Comment (janvorli) [inline src/vm/arm64/stubs.cpp]: In the root clrdefinitions.cmake:
```cmake
if (CLR_CMAKE_PLATFORM_UNIX OR CLR_CMAKE_TARGET_ARCH_ARM64)
  add_definitions(-DFEATURE_IMPLICIT_TLS)
  set(FEATURE_IMPLICIT_TLS 1)
endif(CLR_CMAKE_PLATFORM_UNIX OR CLR_CMAKE_TARGET_ARCH_ARM64)
```
So for ARM64, it is defined for both Windows and Unix.

Comment (sdmaclea) [inline src/vm/arm64/stubs.cpp]: Done.  Closed #12332

**Comment by sdmaclea** [comment]
@jkotas @janvorli This is ready to merge

**Comment by jkotas** [comment]
@jashook Could you please trigger whatever arm64 testing is appropriate?

**Comment by jashook** [comment]
Currently testing is offline and I am looking into the issue. If I cannot find the solution in a reasonable amount of time it is probably worth not blocking.

**Comment by sdmaclea** [comment]
FYI, I did run this through Unix/Arm64 testing locally.

**Comment by sdmaclea** [comment]
@dotnet-bot test Ubuntu arm64 Cross Debug Build

**Comment by sdmaclea** [comment]
ping

**Comment by jashook** [comment]
@dotnet-bot test Windows_NT Arm64 Checked

**Comment by jashook** [comment]
Thank you for waiting @sdmaclea

**Comment by jashook** [comment]
@sdmaclea looks like there is a contract violation

**Comment by sdmaclea** [comment]
@jashook What does a contract violation mean?  I have seen these CONTRACT things floating around, but I haven't figured out the syntax or the tool used to check them.

**Comment by jkotas** [comment]
The CONTRACT things are described in https://github.com/dotnet/coreclr/blob/master/Documentation/coding-guidelines/clr-code-guide.md#2.10

They get checked as asserts at runtime - on Windows only. The contract violation is basically a special form of assertion failure.

I have fixed a contract violation in https://github.com/dotnet/coreclr/pull/12542. Was this change included in the run?

**Comment by jashook** [comment]
It is hard to say, I can restart the job to to see if it repros now.

@dotnet-bot test Windows_NT Arm64 Checked

**Comment by sdmaclea** [comment]
Build timeout
@dotnet-bot test Windows_NT Arm64 Checked

**Comment by sdmaclea** [comment]
@jkotas Contract violation occurs with #12542.  My guess is the CV is in the arm64 tip.  Triggered "Windows_NT arm64 Cross Checked Build and Test" on #12436 to test hypothesis.

**Comment by jashook** [comment]
@sdmaclea Windows arm64 jobs should be clean currently.

**Comment by sdmaclea** [comment]
@jashook @jkotas Confirming #12436 completed w/o contract violation.

**Comment by sdmaclea** [comment]
@dotnet-bot test Ubuntu arm64 Cross Debug Build

**Comment by sdmaclea** [comment]
@jashook FWIW Looks like arm64_debug_small_page_size_prtest pass, but arm64_debug_small_page_size_flow_prtest do not.
+ #12334 passed https://ci.dot.net/job/dotnet_coreclr/job/master/job/arm64_debug_small_page_size_prtest/250/
+ #12333 failed
https://ci.dot.net/job/dotnet_coreclr/job/master/job/arm64_debug_small_page_size_flow_prtest/173/

**Comment by sdmaclea** [comment]
@jkotas I removed the contract line I added.  Since I do not have a Windows Arm64 platform, this the CV is not something I can easily debug.  I'll open an issue to add the contract.

**Comment by sdmaclea** [comment]
@jkotas @dotnet/arm64-contrib please review and/or merge

**Comment by jkotas** [comment]
@dotnet/arm64-contrib Are Windows amd64 tests passing on this change?

I see that `Windows_NT arm64 Cross Debug Build` is green, but I do not see any tests run in the log.

> Remove change to InitJITHelpers1() contract

IIRC, contracts that failed on this earlier were not related to InitJITHelpers1. I do not think that there was anything wrong with the InitJITHelpers1 contract that you have added. It is ok to look into it independently since you have filled issue on it.

**Comment by sdmaclea** [comment]
@dotnet-bot test Windows_NT arm64 Checked

**Comment by sdmaclea** [comment]
@jkotas Since I do not have Windows test platform to debug Contract violation, I disabled the change for Windows and changed #13053 to match.

**Comment by sdmaclea** [comment]
@jkotas 
> I assume that you have done a test pass to verify that there are not regressions on Linux.

Yes.  I have been using the patch for all my runs for the last two months.  Either way, I will try to get someone form the QDT windows team to debug the Contract violation.

> The regression will likely show up as a intermittent crash during GC.

Will GC stress testing help flush it out?

arm64 GC is not perfect yet.  there are still interrmittent failures.  So It may not be obvious that it is a CV related issue.

**Comment by jkotas** [comment]
> I have been using the patch for all my runs for the last two months

This should be good enough.

> Will GC stress testing help flush it out?

Maybe. It is best to debug the known problem on Windows first, I think.

**Comment by sdmaclea** [comment]
@dotnet-bot test Ubuntu arm64 Cross Debug Build

**Comment by sdmaclea** [comment]
Please merge

File: src/vm/arm64/stubs.cpp [hunk 2]
```diff
@@ -1086,6 +1087,31 @@ void JIT_TailCall()
 #if !defined(DACCESS_COMPILE) && !defined(CROSSGEN_COMPILE)
 void InitJITHelpers1()
 {
+#ifdef FEATURE_PAL // TODO
```
Comment (sdmaclea) [inline src/vm/arm64/stubs.cpp]: This should be removed to support Windows

**Pull request merged**